waspc/src/Wasp/Generator/NpmInstall.hs

```diff
@@ -0,0 +1,106 @@
+module Wasp.Generator.NpmInstall
+  ( ensureNpmInstall,
+  )
+where
+
+import Control.Concurrent (Chan, newChan, readChan)
+import Control.Concurrent.Async (concurrently)
+import Control.Monad (when)
+import qualified Data.Aeson as Aeson
+import qualified Data.ByteString.Lazy as B
+import StrongPath (Abs, Dir, File', Path', Rel, relfile, (</>))
+import qualified StrongPath as SP
+import System.Directory (doesFileExist, removeFile)
+import System.Exit (ExitCode (..))
+import Wasp.AppSpec (AppSpec)
+import Wasp.Generator.Common (ProjectRootDir)
+import qualified Wasp.Generator.Job as J
+import Wasp.Generator.Job.IO (printPrefixedJobMessage)
+import Wasp.Generator.Monad (GeneratorError (..), GeneratorWarning (..))
+import qualified Wasp.Generator.NpmDependencies as N
+import Wasp.Generator.ServerGenerator as SG
+import qualified Wasp.Generator.ServerGenerator.Setup as ServerSetup
+import Wasp.Generator.WebAppGenerator as WG
+import qualified Wasp.Generator.WebAppGenerator.Setup as WebAppSetup
+
+-- | Run npm install if needed
+ensureNpmInstall :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO ([GeneratorWarning], [GeneratorError])
+ensureNpmInstall spec dstDir = do
+  let appSpecFullStackPackageJsonDependencies = N.buildFullStackNpmDependencies spec SG.waspNpmDependencies WG.waspNpmDependencies
+  case appSpecFullStackPackageJsonDependencies of
+    Left message -> return ([], [GenericGeneratorError ("npm install failed: " ++ message)])
+    Right fullStackDependencies -> do
+      needed <- isNpmInstallNeeded fullStackDependencies dstDir
+      if needed
+        then executeNpmInstall fullStackDependencies dstDir
+        else return ([], [])
+
+-- Run npm install for desired AppSpec dependencies
```

Comment:
We are making an assumption here, that the app spec deps we are given, are matching the current state of package.json files that were generated on the disk during generation phase, right? Should we write a comment about that?

Reply:
One way to avoid this assumption would be to read package.json files directly from the disk, which is what you suggested before but I advocated on not doing it.
But looking at it now, it would remove this assumption. So maybe we should put a TODO for exploring this as an option in the future?
It wouldn't actually even be very hard, we would just need to read the JSON files. Hm.

Reply:
We'd need to read the JSON files of both packages, extract the dependencies, throw them in a fullStack dependencies structure, and in the end save them. 

Pros:

- We only construct dependencies from app spec once during the generator phase

Cons:

- We save stuff to disk only to load it up again
- 
- What if generation fails due to conflicts? We are left with an older version of `package.json`, which should be okay as we won't try to reinstall it, just like now. But what if generation of the server failed due to conflicts and generation of the webapp succeeds? It would try to rerun `npm install` whereas currently it refuses to do so because there are conflicts in the full stack.
 
That's going to disk to fetch information we just generated to introduce a decoupling between the generator phase and the setup phase we just introduced which is inherently highly coupled anyway, with new a hard to reason about behavior thrown in.

Reply:
Yup, that is ok analysis.

I am not sure what is better.

This is not about coupling really, this is about having an assumption that AppSpec passed in the code is in sync with the package.json files on the disk, which I think is not obvious from the code to somebody stumbling at this piece of code. I am trying to imagine a developer looking at this piece of code trying to figure it out, and I am imagining that figuring out this relationship between in-memory AppSpec and package.json files on disk (which are used at the end by `npm install` command) is not easy to figure out.

That is why I wrote that I think it is good enough for now if we at least document this assumption/relationship well.

Reply:
When does what is generated *not* reflect the AppSpec?